### PR TITLE
video updates

### PR DESCRIPTION
**Pull request by rfomin**
Now the resolution changes in 100px increments, I got good results with 1080p and 1440p displays.

One issue that remains is HOM, which sometimes shows buffer at different resolution, not sure if this is worth fixing. I'm interested in maps where it was used intentionally.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Automap is broken with this change, but drs feels smooth.

**Comment by rfomin** [comment]
> Automap is broken with this change, but drs feels smooth.

It's because of `pitch` here https://github.com/fabiangreffrath/woof/pull/1345/commits/9f0197338f9b024bfc8d87ab8432796259c58062

File: src/i_video.c [hunk 1]
```diff
@@ -494,9 +480,10 @@ static void DynamicResolution(void)
     double actualpercent = actual / target;
 
     #define DRS_MIN_HEIGHT 400
-    #define DRS_DELTA 0.5
+    #define DRS_DELTA 0.1
```
Comment (fabiangreffrath) [inline src/i_video.c:483]: Does this mean that 10% deviation from target frame rate are tolerated?

**Comment by rfomin** [comment]
Yes.

ср, 20 дек. 2023 г., 16:27 Fabian Greffrath ***@***.***>:

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
Hm, the automap is still somewhat broken:

woof0000

Also, it seems that the IDRATE cheat reports the unstretched height dimension, whereas the debug output printed to console reports the stretched ones (i.e. multiplied with factor 1.2). I find this confusing. Could we please agree on one format?

Also, if you say that the resolution is changed in steps of 100p, does this apply to unstretched or stretched vertical dimensions? Because in the latter case, this would mean steps in framebuffer size of 100p/1.2 = 66.7p, which I don't think sounds reasonable.

**Comment by fabiangreffrath** [comment]
Two more things, sorry: Automap works in widescreen mode, but my resolution still resets to 1918x1080 instead of the actual 1920x1080.

woof0001

**Comment by fabiangreffrath** [comment]
There is a 1px high line of game content visible between automap background and status bar.

**Comment by rfomin** [comment]
> Also, it seems that the IDRATE cheat reports the unstretched height dimension, whereas the debug output printed to console reports the stretched ones (i.e. multiplied with factor 1.2). I find this confusing. Could we please agree on one format?

Upstretched is more relevant, I think.

> Also, if you say that the resolution is changed in steps of 100p, does this apply to unstretched or stretched vertical dimensions? Because in the latter case, this would mean steps in framebuffer size of 100p/1.2 = 66.7p, which I don't think sounds reasonable.

Stretching is done on the GPU, almost "free" and with sub-pixel accuracy. I think we don't have to worry about it as long as the image looks "right".

> but my resolution still resets to 1918x1080 instead of the actual 1920x1080.

I don't think we should care about exact numbers in DRS mode, a stable image is much more important. The "IDRATE" cheat is for developers, most users will not use it.
We will still have a slider in the menu or config options where users can set the exact resolution.

This PR still has problems on high refresh rate monitors. I'll try to test it at 144Hz tomorrow.

**Comment by fabiangreffrath** [comment]
I could get width back to 19200 by applying the following simplifications:

```patch
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -990,17 +990,15 @@ static void ResetResolution(int height)
             break;
     }
 
-    double aspect_ratio = (double)w / (double)h;
-
-    if (aspect_ratio > 2.4)
+    if (10 * w > 24 * h)
     {
-        aspect_ratio = 2.4;
+        w = 24;
+        h = 10;
     }
 
-    video.unscaledw = (int)(unscaled_ah * aspect_ratio);
+    video.unscaledw = unscaled_ah * w / h;
 
-    double vertscale = (double)actualheight / (double)unscaled_ah;
-    video.width = (int)(video.unscaledw * vertscale);
+    video.width = actualheight * w / h;
     video.width = (video.width + 1) & ~1;
 
     video.deltaw = (video.unscaledw - NONWIDEWIDTH) / 2;
```

**Comment by rfomin** [comment]
> I could get width back to 1920 by applying the following simplifications:

Sure, this is the original approach. But black bars appear when switching resolutions. Try disabling the menu background and switching to "double"/"triple"/"high" resolution, you will see a slight "movement" of the image. The current approach makes it more stable, see the discussion that starts here: https://github.com/fabiangreffrath/woof/issues/1334#issuecomment-1858933593

I think pretty numbers don't matter, a stable picture is more important. Surely there must be a smarter and more accurate way to do this, happy to try any suggestions.

**Comment by bangstk** [comment]
> I could get width back to 1920 by applying the following simplifications:
So the reason all this is happening is that the 16:9 width for 1x scale is 426.666666666 px...which gets truncated to 426 due to unscaledw being an int.

With your proposed changes, there are now other artifacts, such as:
At Double and Triple, there are some uneven columns that are not 2px or 3px wide (BEFORE any scaling to fit screen is applied), likely due to the difference in ratio between int truncated unscaledw/unscaledah and the real video.width/height causing a non-integer scale. This results in uneven scaling even with 'smooth scaling' being on.

The parts you simplified out were a solution to that which basically enforces that everything is a scaled multiple of the int truncated unscaledw rather than the real fractional 426.66666, so that all pixels remain uniform no matter the scale size.

It does have the side effect of some small gap to the left and right at 16:9, because those are also there in 1x.

The alternative would be figuring out what calculation is resulting in uneven columns at 2x and 3x scale (particularly around the status bar face area) but that is probably pretty tricky.

**Comment by fabiangreffrath** [comment]
Thanks, got the explanation. My hope was that "pretty numbers" were a hint at "correct math" but apparently that's not the case here. Please proceed. 😉

**Comment by liPillON** [comment]
there's no way around it: with integer math there's only so many internal hor.res. which will result in integer screen sizes for all supported aspect ratios.. what about limiting the drs logic to those?

this table sums them up, limiting to thos which are also divisible by 4:

|internal hor.res.	|final 4:3	| final 16:10	| final 16:9	| final 21:9	|
|-------------------|-----------|---------------|---------------|---------------|
|300	            | 480x360	| 576x360	    | 640x360	    | 840x360	    |
|450	            | 720x540	| 864x540	    | 960x540	    |1260x540	    |
|600	            | 960x720	|1152x720	    |1280x720	    |1680x720	    |
|750	            |1200x900	|1440x900	    |1600x900	    |2100x900	    |
|900	            |1440x1080	|1728x1080	    |1920x1080	    |2520x1080	    |
|1200	            |1920x1440	|2304x1440	    |2560x1440	    |3360x1440	    |

**Comment by rfomin** [comment]
> there's no way around it: with integer math there's only so many internal hor.res. which will result in integer screen sizes for all supported aspect ratios.. what about limiting the drs logic to those?

That's what this PR does, but only using 100px steps for the internal resolution. Honestly, both 150px and 100px steps look bad, we may have to use only 200px step. This will make the DRS not gradual, but with except for "linear sky" jumps, there will be virtually no "vibration".

**Comment by liPillON** [comment]
well yes, I was just highlighting that with those 150px steps you'll get resolutions that would not require you to compensate for rounding issues

**Comment by bangstk** [comment]
> well yes, I was just highlighting that with those 150px steps you'll get resolutions that would not require you to compensate for rounding issues

The selection of internal resolution isn't where the rounding issues are coming from, it's the math involved in determining how to draw scaled patches based on an 'original 1x' 16:9 resolution of 426.6666666666x200.

It truncates to 426x200, which is not quite 16:9. When calculating the scaling to get from there to an actual 16:9 resolution (such as 1280x720) the scale for the width ends up just a tad higher than an integer, which is where the uneven columns are coming from.

**Comment by rfomin** [comment]
> I think in particular it's "video.xstep" in V_Init where this is mainly happening.
> If we're settling on screen scaling steps of 200 then that can probably be changed from being fixed to being int based.

But we still want to be able to set any resolution. I tested the current PR on a 1440p 144Hz display, 100px DRS steps looks pretty good at high resolutions. I think if we fix the hysteresis it will be satisfactory.

File: src/i_video.c [hunk 2]
```diff
@@ -494,28 +484,43 @@ static void DynamicResolution(void)
...
+    frame_counter++;
+    averagepercent = (averagepercent + actualpercent) / frame_counter;
+
     if (actualpercent > DRS_GREATER)
     {
         double reduction = (actualpercent - DRS_GREATER ) * 0.2;
         newheight = (int)MAX(DRS_MIN_HEIGHT, oldheight - oldheight * reduction);
     }
-    else if (actualpercent < DRS_LESS)
+    else if (averagepercent < DRS_LESS && frame_counter > targetrefresh)
```
Comment (rfomin) [inline src/i_video.c:503]: @bangstk Now it increases the resolution after about 1 second. A simple approach, but it works well enough for me.

**Comment by fabiangreffrath** [comment]
I think we'll need a comment in the source code that we can refer people to. Something like the comment that I added just recently to explain the choice of horizontal pixels in widescreen mode, which of course is more obsoleted by this approach. 

But, I already know that people will complain and ask for the reasons for the odd-ball resolutions. We'll need something like this, but broken down into a source comment:

https://github.com/fabiangreffrath/woof/pull/1345#issuecomment-1866132187

**Comment by rfomin** [comment]
> But, I already know that people will complain and ask for the reasons for the odd-ball resolutions.

If I remember correctly, people were complaining about black bars on the sides of the screen. Okay, I'll try to add more comments about resolutions and DRS parameters when everything is finalized.

**Comment by fabiangreffrath** [comment]
Thank you! Trust me, there *will* be complaints about everything that we change. Or at least questions. And it's always good to be prepared and be able to point people to some comment in the code and say "see, everything is working as expected". 😉

**Pull request merged**